File: app/src/main/java/com/iogarage/ke/pennywise/PaymentView.java
package com.iogarage.ke.pennywise;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import com.google.android.material.textfield.TextInputLayout;
import androidx.core.view.LayoutInflaterCompat;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.appcompat.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.afollestad.materialdialogs.DialogAction;
import com.afollestad.materialdialogs.MaterialDialog;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdView;
import com.iogarage.ke.pennywise.entities.DaoSession;
import com.iogarage.ke.pennywise.entities.Debt;
import com.iogarage.ke.pennywise.entities.DebtDao;
import com.iogarage.ke.pennywise.entities.Payment;
import com.iogarage.ke.pennywise.entities.PaymentDao;
import com.iogarage.ke.pennywise.entities.Reminder;
import com.iogarage.ke.pennywise.entities.ReminderDao;
import com.iogarage.ke.pennywise.util.Util;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;

import java.util.List;

import javax.inject.Inject;

import autodagger.AutoInjector;
import autodagger.AutoSubcomponent;
import io.multimoon.colorful.CAppCompatActivity;
import uk.co.chrisjenx.calligraphy.CalligraphyContextWrapper;

import static io.multimoon.colorful.ColorfulKt.Colorful;


@AutoSubcomponent
@AutoInjector
@DaggerScope(PaymentView.class)
public class PaymentView extends CAppCompatActivity {
    public static final String TRANSACTION_ID = "param1";

    private static final int PAY = 1;
    private static final int EDIT = 2;

    private long transactionId;
    private Debt transaction;

    @Inject
    DaoSession daoSession;

    private DebtDao transactionDao;

    private TextView phoneNumber;
    private TextView amount;
    private TextView paidAmount;
    private TextView balance;
    private TextView name;
    private TextView note;
    private LinearLayout noteLayout;
    private TextView dueDate;
    private RecyclerView paymentList;
    private Toolbar toolbar;
    private LinearLayout paymentListHeader;

    private PaymentDao paymentDao;
    private String currency;
    private String msg_template;
    private ReminderDao reminderDao;
    private AdView adView;
    private TextView penaltyText;
    private TextView topUpText;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_payment);
        setTitle("Payment History");

        toolbar = findViewById(R.id.toolBar);
        setSupportActionBar(toolbar);


        adView = findViewById(R.id.adView);
        AdRequest adRequest = new AdRequest.Builder()
                .addTestDevice("F76CADC28B22D55BB68E255C034A3996")
                .addTestDevice("825E1DA812A6CB10A5CCB7314FF4387B")
                .build();
        adView.loadAd(adRequest);

        TextView title = findViewById(R.id.title);
        title.setText("");


        phoneNumber = findViewById(R.id.loan_details_phone_no_text);
        amount = findViewById(R.id.loan_details_loanamount);
        paidAmount = findViewById(R.id.loan_details_paid);
        balance = findViewById(R.id.loan_details_balance);
        name = findViewById(R.id.person_name);
        note = findViewById(R.id.note);
        noteLayout = findViewById(R.id.note_layout);
        dueDate = findViewById(R.id.loan_details_due_date);
        paymentList = findViewById(R.id.loansdetailslist);
        penaltyText = findViewById(R.id.loan_penalty_paid);
        topUpText = findViewById(R.id.loan_topup_amount);
        paymentListHeader = findViewById(R.id.loan_payment_list_header);

        ((PennyApp) getApplication())
                .getComponent()
                .plusPaymentViewComponent()
                .inject(this);

        transactionDao = daoSession.getDebtDao();
        paymentDao = daoSession.getPaymentDao();
        reminderDao = daoSession.getReminderDao();

        transactionId = getIntent().getLongExtra(TRANSACTION_ID, 0);
        transaction = transactionDao.loadDeep(transactionId);
    }

    public PaymentDao getPaymentDao() {
        return paymentDao;
    }

    private boolean hasPayment() {
        boolean payments = false;

        if (transaction != null)
            payments = (transaction.getPayments().size() > 0);

        return payments;
    }

    @Override
    protected void attachBaseContext(Context newBase) {
        super.attachBaseContext(CalligraphyContextWrapper.wrap(newBase));
    }

    protected void loadData() {

        if (transaction != null) {

            SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(this);

            if (sharedPref != null) {
                currency = sharedPref.getString(getString(R.string.prefCurrency), "");
                msg_template = sharedPref.getString(getString(R.string.prefTemplate), "");
            }

            if (!currency.isEmpty()) {
                if (currency.equalsIgnoreCase("146"))
                    currency = "";
            }

            amount.setText(Util.formatCurrency(transaction.getAmount()));
            balance.setText(currency + " " + Util.formatCurrency(transaction.getBalance()));

            phoneNumber.setText(transaction.getPhonenumber());
            name.setText(transaction.getPersonname());

            if (transaction.getNote().isEmpty())
                noteLayout.setVisibility(View.GONE);
            else
                noteLayout.setVisibility(View.VISIBLE);

            note.setText(transaction.getNote());

            LinearLayoutManager layoutManager = new LinearLayoutManager(
                    this);
            layoutManager.setOrientation(LinearLayoutManager.VERTICAL);
            paymentList.setLayoutManager(layoutManager);
            paymentList.addItemDecoration(new DividerItemDecoration(this, DividerItemDecoration.VERTICAL_LIST));
            paymentList.setItemAnimator(new DefaultItemAnimator());

            List<Payment> dataset = transaction.getPayments();

            double[] totals = new double[]{0.0, 0.0, 0.0};

            for (Payment payment : dataset) {
                if (payment.getDescription().equalsIgnoreCase(getString(R.string.payment)))
                    totals[0] += payment.getAmountpaid();
                if (payment.getDescription().equalsIgnoreCase(getString(R.string.penalty)))
                    totals[1] += payment.getAmountpaid();
                if (payment.getDescription().equalsIgnoreCase(getString(R.string.topup)))
                    totals[2] += payment.getAmountpaid();

            }


            paidAmount.setText(String.format(getString(R.string.loan_details_place_holder_text), Util.formatCurrency(totals[0])));

            if (totals[1] != 0) {
                penaltyText.setVisibility(View.VISIBLE);
                findViewById(R.id.textView8).setVisibility(View.VISIBLE);
                penaltyText.setText(String.format(getString(R.string.loan_details_place_holder_text), Util.formatCurrency(Math.abs(totals[1]))));
            } else {
                penaltyText.setVisibility(View.GONE);
                findViewById(R.id.textView8).setVisibility(View.GONE);
            }

            if (totals[2] != 0) {
                topUpText.setVisibility(View.VISIBLE);
                findViewById(R.id.textView9).setVisibility(View.VISIBLE);
                topUpText.setText(String.format(getString(R.string.loan_details_place_holder_text), Util.formatCurrency(Math.abs(totals[2]))));
            } else {
                topUpText.setVisibility(View.GONE);
                findViewById(R.id.textView9).setVisibility(View.GONE);
            }

            PaymentAdapter mAdapter = new PaymentAdapter(dataset, this);
            paymentList.setAdapter(mAdapter);

            dueDate.setText("Due date: " + Util.formatDate(transaction.getPaydate()));

        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.pay_menu, menu);

        if (transaction != null) {
            if (transaction.getPaid()) {
                menu.removeItem(R.id.action_pay);
                menu.removeItem(R.id.action_edit);
            }

            if (transaction.getType() == TransactionView.BORROWING)
                menu.removeItem(R.id.action_message);

            if (hasPayment())
                menu.removeItem(R.id.action_edit);
        }
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // handle item selection
        switch (item.getItemId()) {
            case R.id.action_pay:
                payDialog(R.id.action_pay);
                return true;
            case R.id.action_topup:
                payDialog(R.id.action_topup);
                return true;
            case R.id.action_penalty:
                payDialog(R.id.action_penalty);
                return true;
            case R.id.action_edit:
                if (hasPayment())
                    return true;
                Intent intent = new Intent(this, TransactionView.class);
                intent.putExtra("TRANSACTIONID", transactionId);
                startActivity(intent);
                return true;
            case R.id.action_message:
                sendMessage();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void sendMessage() {
        if (transaction != null) {
            if (!msg_template.isEmpty()) {
                msg_template = msg_template.replace("@name", transaction.getPersonname().split(" ")[0]);
                msg_template = msg_template.replace("@amount", currency + " " + Util.formatCurrency(transaction.getBalance()));
            }

            Intent smsIntent = new Intent(android.content.Intent.ACTION_VIEW);
            smsIntent.setType("vnd.android-dir/mms-sms");
            smsIntent.putExtra("address", transaction.getPhonenumber());
            smsIntent.putExtra("sms_body", msg_template);
            startActivity(Intent.createChooser(smsIntent, "Send SMS reminder"));
        }
    }

    private void editTransaction() {
        Intent intent = new Intent(this, TransactionView.class);
        intent.putExtra("TRANSACTIONID", transactionId);
        startActivity(intent);
    }

    @Subscribe
    public void paymentSelected(final PaymentSelected event) {

        //String amount = event.getPayment().getAmountpaid().toString();

        /*MaterialDialog materialDialog = new MaterialDialog.Builder(this)
                .title(R.string.update_payment)
                .content(String.format(getString(R.string.payment_amount), currency, Util.formatCurrency(transaction.getBalance())))
                .inputType(InputType.TYPE_CLASS_NUMBER)
                .positiveText(R.string.update)
                .negativeText(R.string.cancel)
                .neutralText(R.string.delete)
                .onNeutral(new MaterialDialog.SingleButtonCallback() {
                    @Override
                    public void onClick(@NonNull MaterialDialog dialog, @NonNull DialogAction which) {
                        Payment payment = event.getPayment();
                        double original = event.getPayment().getAmountpaid();

                        if (payment.getDescription().equals(getString(R.string.payment)))
                            original *= -1;

                        paymentDao.delete(payment);
                        updateBalance(original);
                        loadData();
                    }
                })
                .input(null, amount, false, new MaterialDialog.InputCallback() {
                    @Override
                    public void onInput(MaterialDialog dialog, CharSequence input) {
                        double amount = Util.parseDouble(input.toString());
                        double original = event.getPayment().getAmountpaid();
                        double difference = amount - original;
                        Payment payment = event.getPayment();
                        payment.setAmountpaid(amount);
                        paymentDao.update(payment);

                        if (!payment.getDescription().equals(getString(R.string.payment)))
                            difference *= -1;

                        updateBalance(difference);
                        loadData();
                    }
                }).build();

        materialDialog.show();*/

        //////////////////////////////////////////////////////////////////////////////////////////////////////////////
        String content = String.format(getString(R.string.payment_amount), currency, Util.formatCurrency(transaction.getBalance()));
        String title = getString(R.string.update_payment);

        boolean wrapInScrollView = true;
        final MaterialDialog materialDialog = new MaterialDialog.Builder(this)
                .title(title)
                .customView(R.layout.payment_dialog, wrapInScrollView)
                .positiveText(R.string.update)
                .neutralText(R.string.delete)
                .negativeText(R.string.cancel)
                .build();
        TextView contentView = materialDialog.getCustomView().findViewById(R.id.content);
        final TextView date = materialDialog.getCustomView().findViewById(R.id.payment_date);
        final TextView amount = materialDialog.getCustomView().findViewById(R.id.payment_amount);
        final TextView note = materialDialog.getCustomView().findViewById(R.id.payment_note);
        contentView.setText(content);
        date.setText(Util.formatDate(event.getPayment().getPaymentdate()));
        amount.setText(event.getPayment().getAmountpaid().toString());
        note.setText(event.getPayment().getNote());
        materialDialog.show();

        materialDialog.getActionButton(DialogAction.NEUTRAL).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                Payment payment = event.getPayment();
                double original = event.getPayment().getAmountpaid();

                if (payment.getDescription().equals(getString(R.string.payment)))
                    original *= -1;

                paymentDao.delete(payment);
                updateBalance(original);
                loadData();
                materialDialog.dismiss();
            }
        });

        materialDialog.getActionButton(DialogAction.POSITIVE).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                String input = amount.getText().toString();

                double amount = Util.parseDouble(input.toString());
                double original = event.getPayment().getAmountpaid();
                double difference = amount - original;
                Payment payment = event.getPayment();
                payment.setAmountpaid(amount);
                paymentDao.update(payment);

                if (!payment.getDescription().equals(getString(R.string.payment)))
                    difference *= -1;

                updateBalance(difference);
                loadData();
                materialDialog.dismiss();
            }
        });
    }

    protected void payDialog(final int action) {
        String content, title, label;

        if (action == R.id.action_penalty) {
            content = String.format(getString(R.string.penalty_amount), currency, Util.formatCurrency(transaction.getBalance()));
            title = getString(R.string.penalty);
            label = getString(R.string.submit);
        } else if (action == R.id.action_topup) {
            content = String.format(getString(R.string.topup_amount), currency, Util.formatCurrency(transaction.getBalance()));
            title = getString(R.string.action_topup);
            label = "Top Up";
        } else {
            content = String.format(getString(R.string.payment_amount), currency, Util.formatCurrency(transaction.getBalance()));
            title = getString(R.string.payment);
            label = getString(R.string.pay);
        }

        boolean wrapInScrollView = true;
        final MaterialDialog materialDialog = new MaterialDialog.Builder(this)
                .title(title)
                .customView(R.layout.payment_dialog, wrapInScrollView)
                .positiveText(label)
                .negativeText(R.string.cancel)
                .build();
        TextView contentView = materialDialog.getCustomView().findViewById(R.id.content);
        final TextView date = materialDialog.getCustomView().findViewById(R.id.payment_date);
        final TextView amount = materialDialog.getCustomView().findViewById(R.id.payment_amount);
        final TextView note = materialDialog.getCustomView().findViewById(R.id.payment_note);
        final TextInputLayout til = (TextInputLayout) materialDialog.findViewById(R.id.amount_input_layout);

        contentView.setText(content);
        materialDialog.show();

        materialDialog.getActionButton(DialogAction.POSITIVE).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                String input = amount.getText().toString();
                String noteStr = note.getText().toString();
                String dateStr = date.getText().toString();

                double amount = Util.parseDouble(input);


                String description;

                if (action == R.id.action_penalty) {
                    amount *= -1;
                    description = getString(R.string.penalty);
                } else if (action == R.id.action_topup) {
                    amount *= -1;
                    description = getString(R.string.topup);
                } else {
                    description = getString(R.string.payment);

                    if (amount > transaction.getBalance()) {
                        til.setError("Payment amount exceeds balance");
                        return;
                    } else
                        til.setError(null);
                }

                Payment payment = new Payment(null, transactionId, Util.parseDate(dateStr),
                        description, noteStr.trim(), Math.abs(amount));
                paymentDao.insert(payment);
                updateBalance(amount);
                loadData();
                materialDialog.dismiss();
            }
        });

    }

    protected void updateReminder() {
        Reminder reminder = transaction.getReminder();

        if (reminder != null) {
            reminder.setActive(false);
            reminderDao.update(reminder);
        }
    }

    protected void updateBalance(double paid) {

        double balance = transaction.getBalance() - paid;

        if (balance <= 0) {
            transaction.setPaid(true);
            updateReminder();
        }

        transaction.setBalance(balance);
        transactionDao.update(transaction);
        transaction.resetPayments();
    }

    @Override
    public void onResume() {
        super.onResume();
        adView.resume();
        loadData();
        EventBus.getDefault().register(this);

        paymentListHeader.setBackgroundColor(Colorful().getPrimaryColor().getColorPack().normal().asInt());

        if (toolbar != null) {
            toolbar.setNavigationIcon(R.drawable.ic_action_back);
            toolbar.setNavigationOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    setResult(RESULT_CANCELED, new Intent());
                    finish();
                }
            });
        }
    }

    @Override
    public void onPause() {
        adView.pause();
        EventBus.getDefault().unregister(this);
        super.onPause();
    }

    @Override
    public void onDestroy() {
        adView.destroy();
        super.onDestroy();
    }

}

File: app/src/main/java/com/iogarage/ke/pennywise/views/payments/PaymentView.kt
package com.iogarage.ke.pennywise.views.payments

import android.content.Intent
import android.graphics.Bitmap
import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import androidx.core.view.MenuHost
import androidx.core.view.MenuProvider
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import coil.load
import coil.transform.CircleCropTransformation
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.WhichButton
import com.afollestad.materialdialogs.actions.getActionButton
import com.afollestad.materialdialogs.customview.customView
import com.afollestad.materialdialogs.customview.getCustomView
import com.google.android.gms.ads.AdRequest
import com.google.android.material.textfield.TextInputLayout
import com.iogarage.ke.pennywise.R
import com.iogarage.ke.pennywise.api.ContactManager
import com.iogarage.ke.pennywise.databinding.FragmentPaymentBinding
import com.iogarage.ke.pennywise.domain.entity.Payment
import com.iogarage.ke.pennywise.domain.entity.ReminderStatus
import com.iogarage.ke.pennywise.domain.entity.Transaction
import com.iogarage.ke.pennywise.domain.entity.TransactionType
import com.iogarage.ke.pennywise.domain.entity.TransactionWithPayments
import com.iogarage.ke.pennywise.util.AppPreferences
import com.iogarage.ke.pennywise.util.asDateString
import com.iogarage.ke.pennywise.util.asString
import com.iogarage.ke.pennywise.util.toCurrency
import com.iogarage.ke.pennywise.util.viewBinding
import com.iogarage.ke.pennywise.views.settings.CurrencyPreference
import dagger.hilt.android.AndroidEntryPoint
import de.coldtea.smplr.smplralarm.smplrAlarmCancel
import kotlinx.coroutines.launch
import java.time.LocalDate
import javax.inject.Inject
import kotlin.math.abs

@AndroidEntryPoint
class PaymentView : Fragment(R.layout.fragment_payment) {

    @Inject
    lateinit var appPreferences: AppPreferences

    private var hasPayments: Boolean = false
    private val binding by viewBinding(FragmentPaymentBinding::bind)
    private val viewModel: PaymentViewModel by viewModels()
    private var transaction: Transaction? = null
    private var currency: String? = null
    private var msg_template: String? = null
    private lateinit var overflowMenu: Menu

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.payments.collect { state ->
                    when (state) {
                        PaymentUiState.Loading -> {

                        }

                        is PaymentUiState.TransactionWithPaymentList -> {
                            updateUi(state.transaction)
                        }
                    }

                }
            }
        }
    }

    private fun updateUi(item: TransactionWithPayments) {

        transaction = item.transaction
        hasPayments = item.payments.isNotEmpty()



        if (transaction?.paid == true) {
            overflowMenu.removeItem(R.id.action_pay)
            overflowMenu.removeItem(R.id.action_edit)
        }
        if (transaction?.type == TransactionType.BORROWING)
            overflowMenu.removeItem(R.id.action_message)
        if (hasPayments)
            overflowMenu.removeItem(R.id.action_edit)

        currency = appPreferences.getString(CurrencyPreference.CURRENCY_SYMBOL, "")
        msg_template = appPreferences.getString(getString(R.string.prefTemplate), "")


        var photo: Bitmap? = null
        try {
            photo = ContactManager.getPhoto(requireContext(), item.transaction.phoneNumber)
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
        if (photo != null) {
            binding.image.load(photo) {
                placeholder(R.drawable.placeholder)
                transformations(CircleCropTransformation())
            }
        }

        binding.loanDetailsLoanAmount.text = transaction?.amount?.toCurrency()
        binding.loanDetailsBalance.text = transaction?.balance?.toCurrency()
        binding.loanDetailsPhoneNoText.text = transaction?.phoneNumber
        binding.personName.text = transaction?.personName

        binding.layoutTransactionNote.noteLayout.visibility =
            if (transaction?.note?.isEmpty() == true)
                View.GONE
            else
                View.VISIBLE

        binding.layoutTransactionNote.note.text = transaction?.note

        val layoutManager = LinearLayoutManager(requireContext())
        layoutManager.orientation = LinearLayoutManager.VERTICAL
        binding.loanDetailsListLayout.loansDetailsList.layoutManager = layoutManager

        val dataset: List<Payment> = item.payments
        val totals = doubleArrayOf(0.0, 0.0, 0.0)
        for (payment in dataset) {
            if (payment.description.equals(
                    getString(R.string.payment),
                    ignoreCase = true
                )
            ) totals[0] += payment.amountPaid
            if (payment.description.equals(
                    getString(R.string.penalty),
                    ignoreCase = true
                )
            ) totals[1] += payment.amountPaid
            if (payment.description.equals(
                    getString(R.string.topup),
                    ignoreCase = true
                )
            ) totals[2] += payment.amountPaid
        }
        binding.loanDetailsPaid.text = totals[0].toCurrency()
        if (totals[1] != 0.0) {
            binding.loanPenaltyPaid.visibility = View.VISIBLE
            binding.textView8.visibility = View.VISIBLE
            binding.loanPenaltyPaid.text = totals[1].toCurrency()
        } else {
            binding.loanPenaltyPaid.visibility = View.GONE
            binding.textView8.visibility = View.GONE
        }
        if (totals[2] != 0.0) {
            binding.loanTopupAmount.visibility = View.VISIBLE
            binding.textView9.visibility = View.VISIBLE
            binding.loanTopupAmount.text = totals[2].toCurrency()
        } else {
            binding.loanTopupAmount.visibility = View.GONE
            binding.textView9.visibility = View.GONE
        }
        val mAdapter = PaymentAdapter(dataset) {
            paymentSelected(it)
        }
        binding.loanDetailsListLayout.loansDetailsList.adapter = mAdapter
        val payDate = transaction?.payDate?.let { LocalDate.ofEpochDay(it).asString() }
        binding.dueDate.text = payDate
    }


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val adRequest = AdRequest.Builder()
            .build()
        binding.adView.loadAd(adRequest)

        // The usage of an interface lets you inject your own implementation
        val menuHost: MenuHost = requireActivity()

        // Add menu items without using the Fragment Menu APIs
        // Note how we can tie the MenuProvider to the viewLifecycleOwner
        // and an optional Lifecycle.State (here, RESUMED) to indicate when
        // the menu should be visible
        menuHost.addMenuProvider(object : MenuProvider {
            override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
                // Add menu items here
                menuInflater.inflate(R.menu.pay_menu, menu)

                overflowMenu = menu

                if (transaction?.paid == true) {
                    menu.removeItem(R.id.action_pay)
                    menu.removeItem(R.id.action_edit)
                }
                if (transaction?.type == TransactionType.BORROWING)
                    menu.removeItem(R.id.action_message)
                if (hasPayments)
                    menu.removeItem(R.id.action_edit)

            }

            override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
                return when (menuItem.itemId) {
                    R.id.action_pay -> {
                        payDialog(R.id.action_pay)
                        true
                    }

                    R.id.action_topup -> {
                        payDialog(R.id.action_topup)
                        true
                    }

                    R.id.action_penalty -> {
                        payDialog(R.id.action_penalty)
                        true
                    }

                    R.id.action_edit -> {
                        editTransaction()
                        true
                    }

                    R.id.action_message -> {
                        sendMessage()
                        true
                    }


                    else -> false
                }

            }
        }, viewLifecycleOwner, Lifecycle.State.RESUMED)
    }

    private fun sendMessage() {
        transaction?.let {
            if (!msg_template!!.isEmpty()) {
                msg_template = msg_template!!.replace(
                    "@name",
                    it.personName.split(" ".toRegex()).dropLastWhile { it.isEmpty() }
                        .toTypedArray().get(0))
                msg_template = msg_template!!.replace(
                    "@amount",
                    currency + " " + it.balance
                )
            }
            val smsIntent = Intent(Intent.ACTION_VIEW)
            smsIntent.type = "vnd.android-dir/mms-sms"
            smsIntent.putExtra("address", it.phoneNumber)
            smsIntent.putExtra("sms_body", msg_template)
            startActivity(Intent.createChooser(smsIntent, "Send SMS reminder"))
        }
    }

    private fun editTransaction() {
        if (hasPayments)
            return
        transaction?.let {
            val direction =
                PaymentViewDirections.actionToTransactionView(it.transactionId)
            findNavController().navigate(direction)
        }
    }

    private fun paymentSelected(payment: Payment) {
        val content = String.format(
            getString(R.string.payment_amount),
            currency,
            transaction?.balance
        )

        val materialDialog: MaterialDialog = MaterialDialog(requireContext())
            .title(R.string.update_payment)
            .customView(R.layout.payment_dialog, scrollable = true)
            .positiveButton(R.string.update)
            .neutralButton(R.string.delete)
            .negativeButton(R.string.cancel)

        val contentView: TextView =
            materialDialog.getCustomView().findViewById(R.id.content)
        val date: TextView =
            materialDialog.getCustomView().findViewById(R.id.payment_date)
        val amount: TextView =
            materialDialog.getCustomView().findViewById(R.id.payment_amount)
        val note: TextView =
            materialDialog.getCustomView().findViewById(R.id.payment_note)
        contentView.text = content
        date.text = payment.paymentDate.asDateString("dd/MM/yyyy")
        amount.text = payment.amountPaid.toString()
        note.text = payment.note
        materialDialog.show()
        materialDialog.getActionButton(WhichButton.NEUTRAL)
            .setOnClickListener {
                var original = payment.amountPaid
                if (payment.description == getString(R.string.payment))
                    original *= -1.0
                viewModel.deletePayment(payment)
                updateBalance(original)
                materialDialog.dismiss()
            }
        materialDialog.getActionButton(WhichButton.POSITIVE)
            .setOnClickListener {
                val input: String = amount.text.toString()
                val amount = input.toDouble()
                val original = payment.amountPaid
                var difference = amount - original
                payment.amountPaid = amount
                viewModel.updatePayment(payment)
                if (payment.description != getString(R.string.payment)) difference *= -1.0
                updateBalance(difference)
                materialDialog.dismiss()
            }
    }

    private fun payDialog(action: Int) {
        val content: String
        val title: String
        val label: String
        when (action) {
            R.id.action_penalty -> {
                content = String.format(
                    getString(R.string.penalty_amount),
                    currency,
                    transaction?.balance
                )
                title = getString(R.string.penalty)
                label = getString(R.string.submit)
            }

            R.id.action_topup -> {
                content = String.format(
                    getString(R.string.topup_amount),
                    currency,
                    transaction?.balance
                )
                title = getString(R.string.action_topup)
                label = "Top Up"
            }

            else -> {
                content = String.format(
                    getString(R.string.payment_amount),
                    currency,
                    transaction?.balance
                )
                title = getString(R.string.payment)
                label = getString(R.string.pay)
            }
        }

        val materialDialog: MaterialDialog = MaterialDialog(requireContext())
            .title(text = title)
            .customView(R.layout.payment_dialog, scrollable = true)
            .positiveButton(text = label) {
                saveData(action, content, it)
            }
            .negativeButton(R.string.cancel) {

            }

        materialDialog.show()
    }

    private fun saveData(action: Int, content: String, materialDialog: MaterialDialog) {
        val contentView: TextView =
            materialDialog.getCustomView().findViewById(R.id.content)
        val date: TextView =
            materialDialog.getCustomView().findViewById(R.id.payment_date)
        val amount: TextView =
            materialDialog.getCustomView().findViewById(R.id.payment_amount)
        val note: TextView =
            materialDialog.getCustomView().findViewById(R.id.payment_note)
        val til: TextInputLayout =
            materialDialog.findViewById(R.id.amount_input_layout) as TextInputLayout
        contentView.text = content


        val input: String = amount.text.toString()
        val noteStr: String = note.text.toString()
        val dateStr: String = date.text.toString()
        var paidAmount = input.toDouble()
        val description: String

        if (action == R.id.action_penalty) {
            paidAmount *= -1.0
            description = getString(R.string.penalty)
        } else if (action == R.id.action_topup) {
            paidAmount *= -1.0
            description = getString(R.string.topup)
        } else {
            description = getString(R.string.payment)
            if (paidAmount > transaction?.balance!!) {
                til.error = "Payment amount exceeds balance"
                return
            } else til.error = null
        }
        val payment = Payment(
            transactionId = transaction?.transactionId!!,
            paymentDate = LocalDate.now().toEpochDay(),
            description = description,
            note = noteStr.trim { it <= ' ' },
            amountPaid = abs(paidAmount)

        )

        viewModel.addPayment(payment)
        updateBalance(paidAmount)
        materialDialog.dismiss()
    }

    private fun updateReminder() {
        transaction?.let {
            smplrAlarmCancel(requireContext()) {
                requestCode { it.alarmId }
            }
        }
    }

    private fun updateBalance(paid: Double) {
        val balance: Double = transaction?.balance?.minus(paid) ?: 0.0
        if (balance <= 0) {
            transaction?.paid = true
            transaction?.reminderStatus = ReminderStatus.OFF
            updateReminder()
        }
        transaction?.balance = balance
        if (transaction != null)
            viewModel.updateTransaction(transaction!!)
    }

    override fun onResume() {
        super.onResume()
        binding.adView.resume()
    }

    override fun onPause() {
        binding.adView.pause()
        super.onPause()
    }

    companion object {
        private const val PAY = 1
        private const val EDIT = 2
    }
}